packages/dev/serializers/src/USD/USDSerializer.ts

```diff
@@ -0,0 +1,710 @@
+            const y: number = 1 - attribute[i + 1];
+            array.push(`(${x.toPrecision(USDExport.Precision)}, ${y.toPrecision(USDExport.Precision)})`);
+        }
+        return array.join(", ");
+    }
+
+    private static _BuildPrimVars(mesh: AbstractMesh) {
+        let output = "";
+        const count = ((mesh.getVerticesData(VertexBuffer.PositionKind) ?? []).length / 3) * 2;
+        const uvList = [
+            mesh.getVerticesData(VertexBuffer.UVKind) ?? [],
+            mesh.getVerticesData(VertexBuffer.UV2Kind) ?? [],
+            mesh.getVerticesData(VertexBuffer.UV3Kind) ?? [],
+            mesh.getVerticesData(VertexBuffer.UV4Kind) ?? [],
+        ];
+        uvList.forEach((uv, id) => {
+            if (uv.length) {
+                output += USDExport._BuildUV(uv, 0, count);
+            }
+        });
+        return output;
+    }
+
+    private static _BuildUV(uv: FloatArray, id: number, count: number) {
+        let output = "";
+        if (uv?.length > 0) {
+            output += `
+            texCoord2f[] primvars:st${id} = [${USDExport._BuildVector2Array(count, uv)}] (
+                interpolation = "vertex"
+            )`;
+        }
+
+        return output;
+    }
+
+    private static _BuildXform(mesh: AbstractMesh, material: Material) {
+        const name = "Object_" + mesh.uniqueId;
+        if (mesh.parent?.name === "__root__") {
+            (mesh.parent as TransformNode).scaling.z = 1;
+        }
+        const mat = mesh.getWorldMatrix();
+        const transform = USDExport._BuildMatrix(mat);
+
+        if (mat.determinant() < 0) {
+            console.warn("BABYLON.USDExport: USDZ does not support negative scales", mesh);
+        }
+
+        return `def Xform "${name}" (
+        prepend references = @./geometries/Geometry_${mesh.uniqueId}.usda@</Geometry>
+        prepend apiSchemas = ["MaterialBindingAPI"]
+    )
+    {
+        matrix4d xformOp:transform = ${transform}
+        uniform token[] xformOpOrder = ["xformOp:transform"]
+        rel material:binding = </Materials/Material_${material.uniqueId}>
+    }
+    `;
+    }
+
+    static _BuildMatrix(matrix: Matrix) {
+        const array = matrix.toArray() as number[];
+        return `( ${USDExport._BuildMatrixRow(array, 0)}, ${USDExport._BuildMatrixRow(array, 4)}, ${USDExport._BuildMatrixRow(array, 8)}, ${USDExport._BuildMatrixRow(
+            array,
+            12
+        )} )`;
+    }
+
+    private static _BuildMatrixRow(array: number[], offset: number) {
+        return `(${array[offset + 0]}, ${array[offset + 1]}, ${array[offset + 2]}, ${array[offset + 3]})`;
+    }
+
+    private static _BuildCamera(camera: Camera) {
+        const name = `${camera.name}_${camera.uniqueId}`;
+        const mat = camera.getWorldMatrix();
+
+        const transform = USDExport._BuildMatrix(mat);
+
+        if (mat.determinant() < 0) {
+            console.warn("BABYLON.USDExport: USDZ does not support negative scales", camera);

```

Comment:
You are using console.warn throughout the PR. would be better to use the Tools.Warn instead because of compatibility with systems in which Babylon can run.
This is related to the entire PR

Reply:
Roger that, Ill try to get this all fixed up this weekend.